TPGSup/devTPG300.db

```diff
@@ -55,18 +43,272 @@ record(mbbi, "$(P)UNITS")
+    field(SCAN, "Passive")
+    field(PREC, "1")
+    field(LOPR, "0")
+    field(HOPR, "9.9")
+    field(FLNK, "$(P)FUNCTION:LOW:SP:CALC")
+}
+
+record(calc, "$(P)FUNCTION:LOW:SP:CALC")
+{
+    field(SCAN, "Passive")
+    field(PREC, "1")
+    field(INPA, "$(P)FUNCTION:LOW:SP")
+    field(INPB, "$(P)FUNCTION:LOW:SP.LOPR")
+    field(INPC, "$(P)FUNCTION:LOW:SP.HOPR")
+    field(INPD, "$(P)FUNCTION:LOW:SP.PREC")
+    field(CALC, "FLOOR(D*10*MIN(MAX(A,B),C))/(D*10)")
+}
+
+record(longin, "$(P)FUNCTION:LOW:E:SP")
+{
+    field(DESC, "Setpoint for function's low threshold exponent")
+    field(SCAN, "Passive")
+    field(LOPR, "-99")
+    field(HOPR, "99")
+    field(FLNK, "$(P)FUNCTION:LOW:E:SP:CALC")
+}
+
+record(calc, "$(P)FUNCTION:LOW:E:SP:CALC")
+{
+    field(SCAN, "Passive")
+    field(INPA, "$(P)FUNCTION:LOW:E:SP")
+    field(INPB, "$(P)FUNCTION:LOW:E:SP.LOPR")
+    field(INPC, "$(P)FUNCTION:LOW:E:SP.HOPR")
+    field(CALC, "MIN(MAX(A,B),C)")
+}
+
+record(ai, "$(P)FUNCTION:HIGH:SP")
+{
+    field(DESC, "Setpoint for function's high threshold")
+    field(SCAN, "Passive")
+    field(PREC, "1")
+    field(LOPR, "0")
+    field(HOPR, "9.9")
+    field(FLNK, "$(P)FUNCTION:HIGH:SP:CALC")
+}
+
+record(calc, "$(P)FUNCTION:HIGH:SP:CALC")
+{
+    field(SCAN, "Passive")
+    field(PREC, "1")
+    field(INPA, "$(P)FUNCTION:HIGH:SP")
+    field(INPB, "$(P)FUNCTION:HIGH:SP.LOPR")
+    field(INPC, "$(P)FUNCTION:HIGH:SP.HOPR")
+    field(INPD, "$(P)FUNCTION:HIGH:SP.PREC")
+    field(CALC, "FLOOR(D*10*MIN(MAX(A,B),C))/(D*10)")
+}
+
+record(longin, "$(P)FUNCTION:HIGH:E:SP")
+{
+    field(DESC, "Setpoint for function's high threshold exponent")
+    field(SCAN, "Passive")
+    field(LOPR, "-99")
+    field(HOPR, "99")
+    field(FLNK, "$(P)FUNCTION:HIGH:E:SP:CALC")
+}
+
+record(calc, "$(P)FUNCTION:HIGH:E:SP:CALC")
+{
+    field(SCAN, "Passive")
+    field(INPA, "$(P)FUNCTION:HIGH:E:SP")
+    field(INPB, "$(P)FUNCTION:HIGH:E:SP.LOPR")
+    field(INPC, "$(P)FUNCTION:HIGH:E:SP.HOPR")
+    field(CALC, "MIN(MAX(A,B),C)")
+}
+
+record(mbbo, "$(P)FUNCTION:ASSIGN:SP")
+{
+    field(DESC, "Setpoint for function's circuit assignment")
+    field(SCAN, "Passive")
+    field(DTYP, "Soft Channel")
+    field(ZRST, "No assignment")
+    field(ONST, "A1")
+    field(TWST, "A2")
+    field(THST, "B1")
+    field(FRST, "B1")
+    field(FVST, "A1 self-monitor")
+    field(SXST, "A2 self-monitor")
+    field(SVST, "B1 self-monitor")
+    field(EIST, "B1 self-monitor")
+    field(OUT, "$(P)FUNCTION:ASSIGN")
+}
+
+record(longin, "$(P)FUNCTION:ASSIGN")
+{
+    field(SCAN, "Passive")
+}
+
+record(stringout, "$(P)FUNCTION:ASSIGN:SP:OUT") 
+{
+    field(DESC, "Sends new function's settings")
+    field(SCAN, "Passive")
+    field(DTYP, "stream")
+    field(OUT,  "@devTPG300.proto setFunction($(P)FUNCTION,$(P)FUNCTION:LOW:SP:CALC,$(P)FUNCTION:LOW:E:SP:CALC,$(P)FUNCTION:HIGH:SP:CALC,$(P)FUNCTION:HIGH:E:SP:CALC,$(P)FUNCTION:ASSIGN) $(PORT)")
```

Comment:
Do you need to repeat `$(P)FUNCTION` on each argument, or could you just send e.g. `LOW:E:SP:CALC` and use  `\$1\$2` in the protocol?